Collection of mostly cherry-pickable changes related to c++11 features

jonesmz opened the pull request:
Each of the commits in this pull request should be separable form the others if you care to do so.

The purpose of this pull request is to upgrade the nzmqt codebase to take advantage of modern c++11 language features.

Please note that Qt 5.7, released 1 year ago, introduced a hard requirement of c++11 support for the compiler in use. 

Each commit has been tested in sequence (building on the previous) using the nzmqt_test program, with GCC 5.4.0.

jonesmz commented on include/nzmqt/impl.hpp:
```diff
@@ -453,9 +453,6 @@ NZMQT_INLINE void PollingZMQSocket::onMessageReceived(const QList<QByteArray>& m
 
 NZMQT_INLINE PollingZMQContext::PollingZMQContext(QObject* parent_, int io_threads_)
     : super(parent_, io_threads_)
-    , m_pollItemsMutex(QMutex::Recursive)
```
Hrmm, sorry about this. Apparently I accidentally included two "commits" worth of changes into one.

The combined commit is still simple enough to review though.

jonnydee commented:
You're great, thanks. I guess Visual Studio compilers <2017 will have problems with constructor forwarding. Anyway, let's see if someone complains. 

But would you mind to NOT initialize members in the header file. I consider this to be implementation detail and would like to let this be "hidden" in the constructor implementations. Additionally, class initialization code stays in one place and is not distributed over header and implementation files.

mwallnoefer commented:
The commit 26f764d could be applied immediately, since it does not depend on C++11.

jonnydee commented:
The problem is not C++11 -- I'm perfectly fine with using C++11 features. But AFAIK a `using SomeClass::SomeClass;` is not supported with older Visual Studio versions. (Only methods can be made visible using this technique with older versions but not constructors). However, as I said, I'll accept that change anyway. So if you moved back member initialization to the constructor implementations again I'd accept your pull request.

jonesmz commented:
I use constructor forwarding with Visual Studio 2015 for work, and a work colleague tells me he's used it with 2012, though he admits that it was far enough back that he might be getting the version number wrong.


Re: Member initialization. I personally disagree with you on that, but I can remove it from the pull request since it really doesn't matter much.

jonesmz commented:
I rewrote the commit

 	Use c++11 style connect, instead of old string matching version			c2e3029

In place and modified the history of my branch.

Should be what you asked for?

jonnydee commented:
Absolutely great, thank you very much :-)

mwallnoefer commented:
Would it be acceptable to revert only 0cfc7041028b1e5b64c742f642bcbe34152f1db0 since it is not recognised by VS2013?
I would like to add that VS2013 is considered _C++11_-enabled by Qt, although the compliance is not complete. But it should be supported as long as Qt supports it.

jonesmz commented:
I have no strong preference.

I was only offering commits to bring jonnydee/nzmqt closer to my version of it.

If you want to revert it, I won't object.

jonnydee commented:
Please feel free to create a pull request reverting that change. I'll accept it.
